### ISSUE TITLE
Add lance as a supported data format to read/write

### ISSUE THREAD

**Issue by mkleinbort-ic** [pola-rs/polars#14452]
### Description

Lance is emerging as a very powerful data format for tabular data and machine learning. 

It should be possible to integrate it to `polars` to better leverage the predicate pushdown capabilities (via a proposed `pl.scan_lance`)

Pros:
- A native integration into `polars` would offer good support for predicate pushdown and projections
- Good for both libraries

Cons:
- `Lance` is not a very popular data format, and it may be too unstable to expose it to the `polars` userbase

Here is a python showcase of the read-write performance - for testing:

```python

def _read_lance(filepath:str, filters:str=None, rowIds:list[int]=None, columns=None, storage_options=storage_options)->pl.DataFrame:
        
    import os 
    import lance

    # Lance is in the process of adding storage_options to their python api
    # at the moment it requires credentials to be set as env variables

    if storage_options:
        for k,v in storage_options.items():
            os.environ[k] = v
       
    # Lance supports filters (for predicate pushdown) or an explicit list of row numbers to fetch        
    assert filters is None or rowIds is None, 'Only one of filters or rowIds can be provided'
        
    if rowIds:
        # Take is extremely fast
        return pl.from_arrow(lance.dataset(filepath).take(indices=rowIds, columns=columns))
   
    elif filters:
       # filters are strings similar to a where clause. Eg "`userId` = 27"
       # I think this is where a pl.scan_lance would add the most value
       return pl.from_arrow(lance.dataset(filepath).scanner(filter=filters, columns=columns).to_table())
        
    else:
       return pl.from_arrow(lance.dataset(filepath).to_table())
        

def _write_lance(self:pl.DataFrame, filepath, mode='overwrite', storage_options=storage_options, index_on:list[str]|None=None, **kwargs)->None:
    '''Write via the lance.write_dataset API'''
        
    import os 
    import lance
 
    if storage_options:
       for k,v in storage_options.items():
            os.environ[k] = v
                
    dataset = lance.write_dataset(self, filepath, mode=mode, **kwargs)

   # Lance supports indexes for very fast look ups (similar to .take) 
   # Not sure Polars should expose this, but it's pretty good for most usecases        
   if index_on:
       for col in index_on:
           dataset.create_scalar_index(column=col,index_type='BTREE',name=col)
   
   return None     
```

**Comment by mkleinbort-ic** [comment]
Note this is related to, but not a duplicate of https://github.com/pola-rs/polars/issues/12389 given this ask is more for the tabular read-write than the vector search.

**Comment by changhiskhan** [comment]
Also, if you have scalar indices setup, filter pushdowns would allow filtered scans to be extremely fast as well. We'd be interested in collaborating on this if there's openness to having an official polars reader/writer for Lance (like we have with Ray / Daft).

**Comment by owenprough-sift** [comment]
Perhaps this would be a good use case for someone to write an [IO plugin](https://docs.pola.rs/user-guide/plugins/io_plugins/)?

**Comment by mkleinbort-wl** [comment]
Circling back to this. I've had a very good experience writing and using my own `.write_lance`/`.read_lance`

It's the only file format I've come across with truly good support for the arrow standard that offers:
- "append" as a write option
- excellent predicate pushdown

I've tested iceberg and delta, but both were too limited in their support for data types.

**Comment by sashankaryal** [comment]
@mkleinbort-wl how did you do it? Can you share? This is sorely missing

**Comment by mkleinbort-wl** [comment]
Hey, sorry - I blanked this.

Take this as a "till it's implemented" placeholder:

```python
def _get_default_storage_options(...) -> dict[str, str]:
    '''Internal implementation detail - returns cloud creds (S3/Azure/etc.)'''
    ...


def get_lance_dataset(filepath, storage_options=None) -> lance.LanceDataset:

    if storage_options is None:
        storage_options = _get_default_storage_options()

    ds = lance.dataset(filepath, storage_options=storage_options)
    return ds


def _read_lance(filepath, columns=None, predicates: str | dict | None = None, storage_options=None, _return_scanner: bool = False) -> pl.DataFrame:

    ds = get_lance_dataset(filepath, storage_options=storage_options)

    scanner = ds.scanner(
        columns=columns,
        filter=generate_where_predicate_from_dict(predicates) if isinstance(predicates, dict) else predicates,
    )

    if _return_scanner:
        return scanner

    return pl.from_arrow(scanner.to_table())


def _write_lance(self: pl.DataFrame, filepath, mode='append', storage_options=None, index_on: tuple[str] | str | None = None, **kwargs) -> None:
    '''Write via the lance.write_dataset API'''

    if storage_options is None:
        storage_options = _get_default_storage_options()

    dataset = lance.write_dataset(self, filepath, mode=mode, storage_options=storage_options, **kwargs)

    if isinstance(index_on, str):
        index_on = [index_on]

    if index_on is not None:
        for col in index_on:
            dataset.create_scalar_index(column=col, index_type='BTREE', name=col)

    return None


def _from_lance(tb: lance.LanceTable) -> pl.DataFrame:
    return pl.from_arrow(tb.to_arrow())


pl.read_lance = _read_lance
pl.DataFrame.write_lance = _write_lance
pl.from_lance = _from_lance
```

Then there is a separate key function `generate_where_predicate_from_dict` - that is a very tiny "dict to predicate" maker - it works for us given it's a familiar pattern in our work, but not sure I recommend it to others:

```python
from collections.abc import Iterable

def _generate_quoted_list(strings: list[str], quote="'") -> str:
    '''Generates a comma separated quoted list of strings.

    Example:
    ========
    _generate_quoted_list(['x', 'y', 'z'])
    >>> "'x', 'y', 'z'"
    '''
    return ', '.join([f"{quote}{s}{quote}" for s in strings])


def _generate_unquoted_list(items: list) -> str:
    '''Generates a comma separated list of items.

    Example:
    ========
    _generate_unquoted_list([1, 2, 3])
    >>> '1, 2, 3'
    '''
    return ', '.join([f"{i}" for i in items])


# WHERE CLAUSES

def _value_is_datelike(val) -> bool:
    import datetime
    import pendulum

    if isinstance(val, (datetime.date, datetime.datetime)):
        return True

    if isinstance(val, str):
        try:
            pendulum.parse(val)
            return True
        except pendulum.exceptions.ParserError:
            pass

    return False


def _generate_where_range_predicate(fieldName: str, value_range: tuple[float | str | None]) -> str:
    '''Returns a WHERE clause for a value to be in a range'''

    assert len(value_range) == 2 and isinstance(value_range, tuple), f'Requires a tuple of two values, not {type(value_range)=}'

    min_val, max_val = value_range

    DATE_FIELD = _value_is_datelike(min_val) or _value_is_datelike(max_val)

    prefix = 'DATE' if DATE_FIELD else ''

    if isinstance(min_val, str) or _value_is_datelike(min_val):
        min_val = f"'{min_val}'"

    if isinstance(max_val, str) or _value_is_datelike(max_val):
        max_val = f"'{max_val}'"

    match value_range:
        case None, None:
            return ''
        case None, _:
            return f'AND `{fieldName}` <= {prefix} {max_val}'
        case _, None:
            return f'AND `{fieldName}` >= {prefix} {min_val}'
        case _, _:
            return f'AND `{fieldName}` BETWEEN {prefix} {min_val} AND {prefix} {max_val}'
        case _:
            raise NotImplementedError()


def _generate_where_in_predicate(fieldName: str, options: list[str] | list[int] | str | int | None) -> str:
    if options is None:
        return ''

    if isinstance(options, int):
        return f'''AND `{fieldName}` = {options}'''

    if isinstance(options, str):
        return f'''AND `{fieldName}` = DATE '{options}' ''' if _value_is_datelike(options) else f'''AND `{fieldName}` = '{options}' '''

    if isinstance(options, Iterable) and isinstance(options[0], str):
        return f'AND `{fieldName}` IN ({_generate_quoted_list(options)})'

    if isinstance(options, Iterable) and isinstance(options[0], int):
        return f'AND `{fieldName}` IN ({_generate_unquoted_list(options)})'

    raise NotImplementedError(f'{options=} of type {type(options)} is not implemented.')


def _generate_where_predicate_from_predicates(predicates: list[str]) -> str:
    return '\nAND '.join(predicates)


def generate_where_predicate_from_dict(filters: dict) -> str:

    clauses = []
    for k, v in filters.items():

        if k != '__OTHER__':

            if isinstance(v, tuple) and len(v) == 2:
                clauses.append(_generate_where_range_predicate(k, v))
            else:

                if (isinstance(v, Iterable) and len(v) == 2) and (not isinstance(v, list)):
                    print(f'WARNING: did you mean to handle {k} as a range?')
                clauses.append(_generate_where_in_predicate(k, v))
        else:
            # `__OTHER__` is an escape hatch for hand-written predicate strings
            clauses.append(_generate_where_predicate_from_predicates(v))

    def strip_vestigial_prefix(s):
        # Each `_generate_where_*` helper prepends "AND " so they compose cleanly;
        # strip it per-clause so the final ' AND '.join produces a well-formed string.
        if s.startswith("AND "):
            return s.replace("AND ", "", 1).strip()
        return s

    return ' AND '.join([f'({strip_vestigial_prefix(c.strip())})' for c in clauses])


def generate_predicate(**kwargs) -> str:
    return generate_where_predicate_from_dict(kwargs)
```

**Comment by jorritsandbrink** [comment]
I created the `polars-lance` package that adds read/write support for Lance datasets to Polars: https://pypi.org/project/polars-lance/. It's using the IO plugin functionality.

**Comment by Xuanwo** [comment]
Update on the PR's desc:

```diff
- Lance is not a very popular data format, and it may be too unstable to expose it to the polars userbase
+ Lance is a very popular data format now, and it has stable file/table/index format specs to be exposed to the polars userbase
```

**Comment by jonasdedden** [comment]
I actually also was working on a Lance <-> Polars integration in parallel, and just now found out that @jorritsandbrink published something already May of this year 🙈 Anyways, here is what I built: https://github.com/jonasdedden/polars-lance (yes, it also has the same name for now 😅)

Note that this is just a small Research project and shall in no form be read as anything remotely production ready. Large parts of it are pretty hacky too.

My solution is as lazy as possible in comparison (no `DataFrame` has to be materialized but it works with `LazyFrame`), on both reading and writing through custom `scan_lance` and `sink_lance` implementations, and thereby should support larger-than-memory workloads with a bunch of Lance-specific extra features on top. It's also Python-only instead of a statically compiled extension, making it possible to exchange whatever `pylance` or `polars` version you want to use (limited by breaking API changes of course).

I left some LLM-based comparison document here: https://github.com/jonasdedden/polars-lance/blob/main/COMPARISON.md

**Comment by jonasdedden** [comment]
FYI, with `polars-cloud` 0.10.0, a new `sink_batches()` was introduced which enabled distributed & lazy *writes* into a Lance dataset too: https://github.com/jonasdedden/polars-lance#polars-cloud

**Comment by jonasdedden** [comment]
I gave my package some thorough cleanups and now published it on PyPI: https://pypi.org/project/polars-pylance/
In order to differentiate from @jorritsandbrink's package, the name is now slightly different: `polars-*py*lance`.

I also worked on extensive benchmarks on a large AWS machine with datasets up to 200GiB that show a strong performance and memory load advantage of my package compared to `polars-lance`: https://github.com/jonasdedden/polars-pylance/blob/main/COMPARISON.md
